Indexing/ClauseCodeTree.cpp

```diff
@@ -258,23 +256,21 @@ void ClauseCodeTree::remove(Clause* cl)
   incTimeStamp();
 
   CodeOp* op=getEntryPoint();
-  firstsInBlocks.push(op);
+  firstsInBlocks->push(op);
   unsigned depth=0;
   for(;;) {
-    RemovingLiteralMatcher* rlm;
-    Recycler::get(rlm);
-    rlm->init(op, lInfos.array(), lInfos.size(), this, &firstsInBlocks);
-    rlms.push(rlm);
+    rlms->push(Recycled<RemovingLiteralMatcher, NoReset>());
+    RemovingLiteralMatcher* rlm = &*rlms->top();
+    rlm->init(op, lInfos.array(), lInfos.size(), this, &*firstsInBlocks);
 
```

Comment:
Around here, I would argue, things are getting pretty ugly. 

Not sure what to do about it, as the key gain overall was to REPLACE `Recycler` by `Recycled`, but could we give it a thought?

The nested pattern declared slightly above seems to be the root of this.

I am getting some headaches trying to figure out what would need to be done, if anything at all, to first get the `rlm` in some form, initialize it and only then push it to `rlms` (as in the original code).

(The `&*` is also confusing, but I am guessing this needed to get a reference, rather than a pointer, to whatever is stored inside the `Recycled` thing?)

Reply:
> (The &* is also confusing, but I am guessing this needed to get a reference, rather than a pointer, to whatever is stored inside the Recycled thing?)


Yes exactly. If we have `t: Recycled<T>` then `*t: T&` and `&*t: T*`. That's needed because of the signature of `RemovingLiteralMatcher::init`

I agreee that this part of the code is not very nice, but it was already not nice before I touched it. I just refactored it to keep the same behaviour. 

> I am getting some headaches trying to figure out what would need to be done, if anything at all, to first get the rlm in some form, initialize it and only then push it to rlms (as in the original code).

So you could do it like this for example.
```
    Recycled<RemovingLiteralMatcher, NoReset> rlm;
    rlm->init(...);
    rlms->push(std::move(rlm));
```
but then we cannot use the reference to rlm later in the code, as `Recycled` has the same ownership semantics as `unique_ptr`, that is, if it's moved the thing it's being moved out of is invalidated (to ensure it's not being double-freed and so.